### PR TITLE
WIP: Feature 0 lights and security fov behaviour

### PR DESCRIPTION
**Pull request by locojoetive**
## Adds prefabs **GlobalLight**
**GlobalLight** is a game object carrying a global light source. It is necessary for cameras which have light effects turned on, in order to render any Sprites.

## Adds prefabs **Lamp**
**Lamp** is a free-form light source, which can be edited by pressing the "Edit shape" button at the bottom of its **Light2D (Script) Component**. It is considered as the new layer type "Light".
Since the **Field Of View** prefab is created with respect to light inside the scene, the areas lightend-up by Lamp need to be detectable from each field of view. Therefore the attached script **Lamp.cs** creates a trigger-collider during the creation of this object.

## Adds prefab **FieldOfView** for simulating an _effective_ and _in-light field of view_
The Field of View game object has the following scripts attached:
1. FieldOfView.cs
2. FieldOfViewMovement.cs
3. InterpolateMaterial.cs

### 1. FieldOfView.cs
####1. LayerMask WhatBlocksRay:
Determines which layer can block the view.
####2. LayerMask WhatIsSuspicious:
Determines which layer will turn Detected-Mode on, when in view distance and range.
####3. LayerMask WhatIsLight:
Determines which layer of objects will be considered as light source.
####6. bool Detected:
Is true if Detected-Mode is on.
####4. float NormalViewDistance:
Determines the distance of the effective Field of View, It is also present outside a light source.
####5. float InLightViewDistance:
Determines the distance of the In-Light-Field-of-View, It is only present if a light source is in view range.
####6. float FovRange:
Determines the angle of the field of view's cone shape.

### 2. FieldOfViewMovement.cs
By using its coroutine
`public IEnumerator RotateFromAngleToAngleForSecondsAfterSeconds(
   Vector3 fromEulerAngles, 
   Vector3 toEulerAngles, 
   float forSeconds, 
   float afterSeconds
)`
it can rotates the field of view from fromEulerAngles to toEulerAngles in forSeconds seconds after afterSeconds seconds.
The method can be used like this:

`
FieldOfVieMovement fovMovement = GetComponent<FieldOfView>();
... 
StartCoroutine(fovMovement.RotateFromAngleToAngleForSecondsAfterSeconds(
   new Vector3(0,180,0), 
   new Vector3(0,90,45),
   0.5f,
   2f
));
... 
`
After waiting for 2 seconds, this would rotate the field of view linearly from the rotation (0, 180, 0) to (0, 90, 45) in 0.5 seconds.

### 3. InterpolateMaterial.cs

###Notes: 
1. All cameras will have turned on light effects. Therefore all SpriteRenderer components need the material **Sprite-Default-Unlit**
2. Make sure that all SpriteRenderer in your scene have set **Sprite-Default-Unlit** as Material
3. Add the prefab **GlobalLight** to your scene

### CODE REVIEW

File: Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        }
+    }
+    
+    private void DrawRays()
+    {
+        for (int i = 0; i < numberRays; i++)
+        {
+            Vector2 origin = transform.position;
+            Vector2 direction = transform.TransformDirection(detectionRays[i]);
+            RaycastHit2D normalRayHitsPlayer = Physics2D.Raycast(origin, direction, fNormalRayDepth, whatIsPlayer.value);
```
Comment (locojoetive) [inline Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs]: Was ist 'whatIsPlayer'?

Comment (locojoetive) [reply Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs]: Das ist ein Layer der den Spieler identifiziert (Solange er nicht versteckt ist)

Comment (CobbieCobbie) [reply Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs]: ...? Tust du dich selbst reviewen oder was geht?

Comment (CobbieCobbie) [reply Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs]: :D

**Review by CobbieCobbie — COMMENTED** [review]
Woher kommen 103 changed files? Ist das eine geänderte Ordnerstruktur? Hast du in 100 Dateien gecoded?

File: Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs (hunk 1 above)
Comment (locojoetive) [reply Assets/_Scripts/GameMechanic/Enemies/Patrol/SecurityWatch.cs]: 🤣🤣🤣🤣  Ne, hab Henk mal gezeigt was das kann 👯

**Comment by locojoetive** [comment]
> 
> 
> Woher kommen 103 changed files? Ist das eine geänderte Ordnerstruktur? Hast du in 100 Dateien gecoded?

Nope, hab die Struktur verändert.
Meinst du ich hätte das in nem extra branch machen, und den dann hier reinziehen sollen?

**Review by CobbieCobbie — APPROVED** [review]

**Pull request merged**